### PR TITLE
shell: add shell package and modify prep and daemon to use shell

### PR DESCRIPTION
**Pull request by kardianos**
Need a follow-up commit to set the shell from the config.

### CODE REVIEW

**Comment by cortesi** [comment]
Hey there. Thank you - this looks like a good start. I think it can be just a touch simpler, so I'm putting some comments in the code to that effect. Have a look - I'll probably merge and adjust myself in the next day or so if you don't beat me to it. 

I'm mulling over the config file syntax. We could just re-use the variable format like so:

    @shell = bash

Alternatively, we could use a new prefix (a prefix is necessary, unfortunately, because of the freeform path pattern specifications):

    !shell = bash

Thoughts?

File: shell/shell.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	shells     = make(map[string]Interface)
+)
+
+func init() {
+	Register(&Raw{})
+	Register(&Bash{})
+}
+
+// Register a new shell interface.
+func Register(i Interface) {
```
Comment (cortesi) [inline shell/shell.go]: I think we can do without the Register function and the invocations in init. We'll have a small, fixed number of shells (bash, raw, cmd... maybe powershell?), so we can just enter them into the table in the var declaration.

File: shell/shell.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	if _, has := shells[name]; has {
+		panic("shell interface " + name + " already exists")
+	}
+	shells[name] = i
+}
+
+// Command returns a *Cmd. If method is empty then the default shell
+// interface method is used. The line should contain the exec line.
+func Command(method, line string) (*exec.Cmd, error) {
+	shellsLock.Lock()
```
Comment (cortesi) [inline shell/shell.go]: Why this lock?

**Comment by cortesi** [comment]
One more thing - we'll need a method like this: 
  
    Has(method string) bool

This will let us check that a shell in fact exists without invoking anything, so that we can error early if the user is using a modd.conf their system doesn't support.

File: shell/shell.go (hunk 1 above)
Comment (kardianos) [inline shell/shell.go]: By keeping a register function (private or public) you can then add a file like "shell/cmd_windows.go" that registers itself when compiled.

File: shell/shell.go (hunk 2 above)
Comment (kardianos) [inline shell/shell.go]: We could also make Register private and ensure we only call in init functions.

**Comment by kardianos** [comment]
Has(string) bool
Sounds good.

RE variable: That I wasn't sure on. I would be fine with @shell, !shell or @@shell = foo. I also wasn't sure how to best expose variable to the two functions that need it.

If you're alright with keeping register but making it private, I'll also remove the lock.

**Comment by kardianos** [comment]
I think this last commit satisfies everything. Uses "@shell" to set the shell interface, has early detection for verifying shell, and can be expanded later to use OS specific shell interfaces.

I might consider renaming "raw" to "direct" or "exec". I like "exec". Thoughts?

**Comment by cortesi** [comment]
This looks great. There's one more thing to cope with - the argument escaping is now shell-specific. We'll have one quoting strategy for bash/sh, one for cmd.exe and a different one (no escaping) for raw. So we need to add another method to the Interface interface (heh) that looks like "Quote(string) string" (replacing varcmd.quotePath). Then the nicest thing might well be to teach the varcmd module to know about the shell module.

As for the "raw" name, I agree - exec is better.

**Comment by kardianos** [comment]
I'm not convinced that is the case. I'm using "shlex" to take a single string to break it into each argument:
https://github.com/google/shlex/blob/master/shlex_test.go
As you can see from the shlex test, it tries to follow bash like escaping rules.

If we want consistent escaping rules for everything, including future shells, we could use shlex to escape them then use a system specific args to re-encode them. Either way I don't think another method is needed as "shell.Command" returns a *exec.Cmd, which has everything already escaped into arguments.

**Comment by cortesi** [comment]
Ah, ok - I guess a detour via a shell-escaped string is not so bad. Ok, I'll merge this, and we can change the variable name and add documentation in master.

**Comment by cortesi** [comment]
Feh. You slipped in with the exec change already. Thanks. :)

**Pull request merged**